source/shared/cpp/ObjectModel/TextBlockText.cpp

```diff
@@ -0,0 +1,243 @@
+        {
+            // match for long/short/compact
+            bool formatHasSpace = matches[Format].str()[1] == ' ';
+            int formatStartIndex = formatHasSpace ? 2 : 1;
+            formatStyle = matches[Format].str()[formatStartIndex];
+        }
+
+        AddTextSection(matches.prefix().str(), TextSectionFormat::RegularString);
+
+        if (!isDate && formatStyle)
+        {
+            AddTextSection(matches[0].str(), TextSectionFormat::RegularString);
+            text = matches.suffix().str();
+            continue;
+        }
+
+        for (unsigned int idx = 0; idx < indexer.size(); idx++)
+        {
+            if (matches[indexer[idx]].matched)
+            {
+                // get indexes for time attributes to index into conrresponding matches
+                // and covert it to string
+                *addrs[idx] = stoi(matches[indexer[idx]]);
+            }
+        }
+
+        // check for date and time validation
+        if (IsValidTimeAndDate(parsedTm, hours, minutes))
+        {
+            // maches offset sign, 
+            // Z == UTC, 
+            // + == time added from UTC
+            // - == time subtracted from UTC
+            if (matches[TimeZone].matched)
+            {
+                // converts to seconds
+                hours *= 3600;
+                minutes *= 60;
+                offset = hours + minutes;
+
+                wchar_t zone = matches[TimeZone].str()[0];
+                // time zone offset calculation 
+                if (zone == '+')
+                {
+                    offset *= -1;
+                }
+            }
+
+            // measured from year 1900
+            parsedTm.tm_year -= 1900;
+            parsedTm.tm_mon -= 1;
+
+            time_t utc{};
+            // converts to ticks in UTC
+            utc = mktime(&parsedTm);
+            if (utc == -1)
+            {
+                AddTextSection(matches[0], TextSectionFormat::RegularString);
+            }
+
+            wchar_t tzOffsetBuff[6]{};
+            // gets local time zone offset
+            wcsftime(tzOffsetBuff, 6, L"%z", &parsedTm);
+            std::wstring localTimeZoneOffsetStr(tzOffsetBuff);
+            int nTzOffset = std::stoi(localTimeZoneOffsetStr);
+            offset += ((nTzOffset / 100) * 3600 + (nTzOffset % 100) * 60);
+            // add offset to utc
+            utc += offset;
+            struct tm result{};
+
+            // converts to local time from utc
+            if (!LOCALTIME(&result, &utc))
+            {
+                // localtime() set dst, put_time adjusts time accordingly which is not what we want since 
+                // we have already taken cared of it in our calculation
+                if (result.tm_isdst == 1)
+                {
+                    result.tm_hour -= 1;
+                }
+
+                if (isDate)
+                {
+                    std::string plainDate = std::to_string(result.tm_mon) + dateDelimiter +
```

Comment:
consider using std::put_time(&result, "%Ex").

Reply:
in the case of c++ the compact date can be localized (I guess it should change the month/day/year order) but I will test it

---
In reply to: [168270900](https://github.com/Microsoft/AdaptiveCards/pull/1102#discussion_r168270900) [](ancestors = 168270900)

Reply:
Yes, the %Ex can be localized, for example, in English the date February 14th, 2018 is represented as 02/14/2018 in Spanish it's represented as 14/02/2018